In [ ]:
import pandas as pd
import ast
from ragas import EvaluationDataset

# 讀取 CSV
df = pd.read_csv("ragas_eval_with_response_v3.csv", encoding="utf-8-sig")

# 還原 list 欄位
df["retrieved_contexts"] = df["retrieved_contexts"].apply(ast.literal_eval)

# 欄位重新命名
ragas_df = df[["user_input", "retrieved_contexts", "response", "reference"]]

rag_results = EvaluationDataset.from_pandas(ragas_df)
ragas_df

In [5]:
from ragas import evaluate
from ragas.metrics import (
    AnswerCorrectness,
    ContextPrecision,
    ContextRecall,
    Faithfulness,
    AnswerRelevancy,
)
import openai
from ragas.llms import llm_factory
from ragas.embeddings import embedding_factory
from ragas.embeddings import GoogleEmbeddings
from google import genai
from dotenv import load_dotenv
from ragas.embeddings import LangchainEmbeddingsWrapper
from langchain_openai import OpenAIEmbeddings

load_dotenv()

import pandas as pd
import ast
from ragas import EvaluationDataset

# 讀取 CSV
df = pd.read_csv("ragas_eval_with_response_2026-05-22.csv", encoding="utf-8-sig")

# 還原 list 欄位
df["retrieved_contexts"] = df["retrieved_contexts"].apply(ast.literal_eval)

# 欄位重新命名
ragas_df = df[["user_input", "retrieved_contexts", "response", "reference"]]

rag_results = EvaluationDataset.from_pandas(ragas_df)


# 用 AsyncOpenAI
async_client = openai.AsyncOpenAI()

evaluator_llm = llm_factory(
    "gpt-4o-mini",
    provider="openai",
    client=async_client,
    max_tokens=4096,
    max_retries=6,
    timeout=120,  
)
ragas_embeddings = LangchainEmbeddingsWrapper(
    OpenAIEmbeddings(model="text-embedding-3-small")
)

answer_relevancy = AnswerRelevancy(llm=evaluator_llm, embeddings=ragas_embeddings)
answer_relevancy.question_generation.instruction = """根據給定的答案，反推出最有可能對應的問題，並判斷該答案是否為模糊回答。
如果答案是模糊的，noncommittal 給 1；如果答案是明確的，noncommittal 給 0。
模糊回答是指那些迴避、含糊或不明確的回答。
例如，「我不知道」或「我不確定」都屬於模糊回答。
請用繁體中文產生問題。"""

scores = evaluate(
    dataset=rag_results,
    metrics=[
        Faithfulness(llm=evaluator_llm),
        ContextPrecision(llm=evaluator_llm),
        ContextRecall(llm=evaluator_llm),
        answer_relevancy,
        # AnswerCorrectness(llm=evaluator_llm),
    ],
)


print(scores)

C:\Users\User\AppData\Local\Temp\ipykernel_15160\1994392570.py:2: DeprecationWarning: Importing AnswerCorrectness from 'ragas.metrics' is deprecated and will be removed in v1.0. Please use 'ragas.metrics.collections' instead. Example: from ragas.metrics.collections import AnswerCorrectness
  from ragas.metrics import (
C:\Users\User\AppData\Local\Temp\ipykernel_15160\1994392570.py:2: DeprecationWarning: Importing ContextPrecision from 'ragas.metrics' is deprecated and will be removed in v1.0. Please use 'ragas.metrics.collections' instead. Example: from ragas.metrics.collections import ContextPrecision
  from ragas.metrics import (
C:\Users\User\AppData\Local\Temp\ipykernel_15160\1994392570.py:2: DeprecationWarning: Importing ContextRecall from 'ragas.metrics' is deprecated and will be removed in v1.0. Please use 'ragas.metrics.collections' instead. Example: from ragas.metrics.collections import ContextRecall
  from ragas.metrics import (
C:\Users\User\AppData\Local\Temp\ipykernel_1516


statement: 蘇彥碩在使用Git時遇到的最大問題是，有人不會使用Git。
verdict: 1
reason: The context explicitly mentions that one of the major problems is that some people do not know how to use Git, which directly relates to 蘇彥碩's experience.


Evaluating:  10%|▉         | 19/200 [00:10<01:13,  2.45it/s]


statement: 在選擇SDGs主題時，需遵循以下原則。
verdict: 1
reason: The context outlines principles for selecting topics, including the SDGs, which supports this statement.

statement: 題目須排除過去三年的題目。
verdict: 1
reason: The context explicitly states that topics must exclude those from the past three years.

statement: 題目選擇應依據TronClass公告為準。
verdict: 1
reason: The context mentions that topic selection should be based on TronClass announcements.

statement: 應盡量選擇可以找到實際使用者進行訪談的題目。
verdict: 1
reason: The context advises to choose topics that allow for interviews with actual users.

statement: 根據SDGs，應挑選一個有興趣的SDG。
verdict: 1
reason: The context instructs to select an interesting SDG based on the SDGs.

statement: 從挑選的SDG中尋找主題。
verdict: 1
reason: The context indicates that one should find a topic from the selected SDG.


LLM returned 1 generations instead of requested 3. Proceeding with 1 generations.
Evaluating:  14%|█▍        | 28/200 [00:14<00:57,  2.97it/s]


statement: 吳建豐使用Firebase作為資料庫。
verdict: 1
reason: 在上下文中明確提到吳建豐的經驗中使用Firebase作為資料庫，因此這個陳述可以直接推斷。

statement: 吳建豐使用Flask開發後端。
verdict: 1
reason: 上下文中明確提到吳建豐的經驗中使用Flask作為後端開發工具，因此這個陳述可以直接推斷。

statement: 董書妤在開發系統時遇到了使用 Firebase 和 React 這兩種語言操作不熟悉的問題。
verdict: 1
reason: 根據上下文，學長姐在製作系統時初次嘗試使用 Firebase 和 React，並且對操作尚未完全熟悉，因此這一陳述可以直接推斷。

statement: 在開發記帳功能時，資料庫的讀取量超過負荷，最終導致資料庫停止運作。
verdict: 1
reason: 上下文中提到在開發記帳功能時遇到了一些問題，導致資料庫的讀取量超過負荷，這一陳述可以直接推斷。

statement: 這主要是因為 Firebase 的免費版本有讀取量的限制。
verdict: 1
reason: 上下文中明確提到出現資料庫停止運作的問題主要是因為 Firebase 的免費版本有讀取量的限制，因此這一陳述可以直接推斷。

statement: 程式碼中的錯誤導致 React 的 `useEffect` 機制在每次讀取後不斷觸發資料的讀寫操作。
verdict: 1
reason: 上下文中提到在學長姐撰寫的程式碼中存在錯誤，導致 React 的 `useEffect` 機制不斷觸發資料的讀寫操作，因此這一陳述可以直接推斷。

statement: 這種情況造成資料庫超載。
verdict: 1
reason: 上下文中提到資料庫的讀取量超過負荷，最終使得資料庫停止運作，因此這一陳述可以直接推斷。

statement: 董書妤認為在開始操作前應更深入了解系統架構。
verdict: 1
reason: 上下文中提到學長姐認為在開始操作前應更深入了解系統架構，因此這一陳述可以直接推斷。

statement: 董書妤認為在發現錯誤時應逐一排除以進行除錯。
verdict: 1
reason: 上下文中提到學長姐建議在發現錯誤時逐一排除以進行除錯，因此這一陳述可以直接推斷

LLM returned 1 generations instead of requested 3. Proceeding with 1 generations.
Evaluating:  15%|█▌        | 30/200 [00:15<01:08,  2.49it/s]LLM returned 1 generations instead of requested 3. Proceeding with 1 generations.



statement: 在學習 User Story 的時候，最常遇到的問題是對這個概念非常陌生。
verdict: 1
reason: 根據吳建豐的經驗，剛開始學到 user story 這個概念時，確實會感到陌生，因此這個陳述可以直接推斷。

statement: 剛開始學習 User Story 不容易掌握。
verdict: 1
reason: 根據上下文，剛開始對 user story 的概念不熟悉是正常的，因此這個陳述是可以推斷的。

statement: 特別是在如何從使用者的角度思考需求上，會遇到困難。
verdict: 1
reason: 上下文提到剛開始學習時會有困難，特別是在使用者角度思考上，因此這個陳述是合理的。

statement: 根據吳建豐的經驗，一開始可能無法一次就把 User Story 的表格設定好。
verdict: 1
reason: 上下文明確提到剛開始對 user story 不熟悉，因此無法一次設定好表格是可以推斷的。

statement: 需要跟著課程進度討論並進行多次調整才能進入狀況。
verdict: 1
reason: 上下文中提到需要跟著課程進度討論並進行調整，因此這個陳述是正確的。

statement: 這種情況是非常正常的。
verdict: 1
reason: 上下文提到對 user story 不熟悉是正常的，因此這個陳述是可以推斷的。

statement: 因此建議多多練習與討論。
verdict: 0
reason: 雖然上下文沒有直接提到建議多多練習與討論，但從學習的過程中可以推斷出這是合理的建議，因此這個陳述的推斷不夠直接。

statement: 隨著課程的進行，學習者會逐漸適應。
verdict: 1
reason: 上下文提到隨著課程的進行，學習者能夠進入狀況，因此這個陳述是可以推斷的。


Evaluating:  18%|█▊        | 37/200 [00:17<00:54,  3.02it/s]


statement: 根據謝承祐的經驗，git可以用來追蹤組員的進度。
verdict: 1
reason: The context explicitly states that git can be used to track everyone's progress according to 謝承祐's experience.

statement: 每次上傳進度時，git會明確顯示每個人做了什麼。
verdict: 1
reason: The context mentions that every time progress is uploaded, git clearly shows what everyone has done.

statement: 使用git可以自行輸入註記。
verdict: 1
reason: The context states that users can input notes themselves when using git.

statement: 使用git不僅能看到每個成員的工作進展。
verdict: 0
reason: The context does not provide information that git allows seeing each member's work progress beyond tracking, making this statement unverifiable.

statement: 使用git還能幫助組員之間了解整體的開發進度。
verdict: 0
reason: The context does not explicitly state that git helps team members understand the overall development progress, so this cannot be directly inferred.


Evaluating:  22%|██▏       | 43/200 [00:19<00:54,  2.87it/s]


statement: 根據辛語柔的經驗，選擇舊的語言技術主要是因為這門課只有一個學期。
verdict: 1
reason: 辛語柔提到選擇舊技術的原因是因為這門課只有一個學期，因此這個陳述可以直接推斷。

statement: 這門課的時間非常有限。
verdict: 1
reason: 根據辛語柔的經驗，提到這門課只有一個學期，暗示了時間的有限性，因此這個陳述可以推斷。

statement: 如果選擇新的技術，學生們可能會面臨學習新知識的壓力。
verdict: 1
reason: 辛語柔提到如果學習新技術，時間太短會導致壓力暴增，因此這個陳述可以推斷。

statement: 學習新知識的壓力可能導致系統的完整性受到影響。
verdict: 1
reason: 辛語柔提到有些組別因為學習新技術而導致系統功能不完整，因此這個陳述可以推斷。

statement: 辛語柔提到，許多人在學習新的技術上花費了過多時間。
verdict: 1
reason: 辛語柔提到有些人從資料庫到語言全部都學新的，並且花了時間去學，因此這個陳述可以推斷。

statement: 花費過多時間學習新的技術的結果是系統的功能沒有做到完善。
verdict: 1
reason: 辛語柔提到因為花時間學習新技術，系統的功能沒有做到完整，因此這個陳述可以推斷。

statement: 因此，建議在專題時再嘗試新技術。
verdict: 1
reason: 辛語柔提到如果要學新技術，建議在專題時再去學，因此這個陳述可以推斷。

statement: 建議集中學習一種技術，以確保項目的完成度。
verdict: 1
reason: 辛語柔提到只學一個技術可以確保系統的完整性，因此這個陳述可以推斷。


LLM returned 1 generations instead of requested 3. Proceeding with 1 generations.
Evaluating:  24%|██▎       | 47/200 [00:22<01:25,  1.78it/s]LLM returned 1 generations instead of requested 3. Proceeding with 1 generations.



statement: 根據蔡易辰的經驗，蔡易辰在使用 git 的時候遇到的問題是，組內除了蔡易辰以外的其他成員都不太會用 git。
verdict: 1
reason: The context explicitly states that besides 蔡易辰, the other members of the group are not very familiar with using git.

statement: 因此，蔡易辰只能一個一個教其他成員使用 git。
verdict: 1
reason: The context mentions that 蔡易辰 has to teach each member individually due to their lack of knowledge about git.

statement: 蔡易辰提到其他成員對於使用 git 感到害怕。
verdict: 1
reason: The context indicates that the other members are afraid of making mistakes or conflicts when using git.

statement: 其他成員擔心會出錯或衝突。
verdict: 1
reason: The context clearly states that the other members are afraid of errors or conflicts when using git.


Evaluating:  24%|██▍       | 49/200 [00:25<02:21,  1.07it/s]


statement: 在 Sprint Planning 中，Git 主要用於整合程式碼和版本管理。
verdict: 1
reason: The context explicitly mentions that Git is used for code integration and version management during the development process.

statement: 團隊可以利用 Git 來管理不同成員所開發的功能。
verdict: 1
reason: The context implies that Git is used to manage code, which includes features developed by different team members.

statement: 每位成員在各自的分支上進行開發。
verdict: 1
reason: The context states that each member should work on their own branches during development.

statement: 完成後再在會議中將變更合併到主分支。
verdict: 1
reason: The context describes that changes should be merged into the main branch during meetings after development is completed.

statement: 這樣的流程可以有效地避免程式碼衝突。
verdict: 1
reason: The context suggests that this process helps to avoid code conflicts, as it mentions the importance of not having many people working on the same file.

statement: 這樣的流程同時保證整體專案的穩定性。
verdict: 0
reason: While the context emphasizes avoiding conflicts, it does not explicitly 

LLM returned 1 generations instead of requested 3. Proceeding with 1 generations.
Evaluating:  28%|██▊       | 55/200 [00:27<01:02,  2.32it/s]


statement: 在系統開發中，遇到的問題主要包括設計共識不足、文件與實作的落差以及組內合作與溝通。
verdict: 1
reason: 根據多位學姐的經驗，提到了設計共識不足、文件與實作的落差以及組內合作與溝通等問題，因此這個陳述可以直接推斷。

statement: 設計共識不足指的是各組員對於系統設計的想法可能存在差異。
verdict: 1
reason: 文中提到由於各人對於設計的預設想法可能存在差異，因此可以推斷設計共識不足是指各組員對於系統設計的想法存在差異。

statement: 設計共識不足可能導致討論不夠明確或意見不一致。
verdict: 1
reason: 文中提到討論不夠明確或意見不一致時可能需要重新設計，這表明設計共識不足會導致這些問題，因此可以推斷。

statement: 設計共識不足可能需要重新設計。
verdict: 1
reason: 文中明確提到當討論不夠明確或意見不一致時，可能需要重新設計，因此這個陳述可以直接推斷。

statement: 董書妤和周佳欣的經驗中提到設計共識的重要性。
verdict: 1
reason: 文中提到學姐們指出設計共識十分重要，因此這個陳述可以直接推斷。

statement: 文件與實作的落差是指老師希望學生先撰寫文件，以確保實作時能夠依照文件進行。
verdict: 1
reason: 文中提到老師希望學生先寫文件，這正是文件與實作的落差的定義，因此可以推斷。

statement: 學生在實作過程中常常會遇到時間不足的問題。
verdict: 1
reason: 文中提到時間不夠可能導致某個功能做特別久，因此可以推斷學生在實作過程中會遇到時間不足的問題。

statement: 學生在開發過程中會想新增某些功能，而文件已經撰寫完畢。
verdict: 1
reason: 文中提到在實作到一半可能想加某個功能，但文件已經打好了，因此可以推斷這個陳述是正確的。

statement: 文件與實作的落差導致需要花額外時間修改文件。
verdict: 1
reason: 文中提到如果實作到一半想加某個功能，會需要花額外的時間修改文件，因此這個陳述可以直接推斷。

statement: 蔡易辰同學提及文件與實作的落差問題。
verdict: 1
reason: 文中明確提到蔡易辰的經驗中

Evaluating:  29%|██▉       | 58/200 [00:31<02:11,  1.08it/s]LLM returned 1 generations instead of requested 3. Proceeding with 1 generations.



statement: 蔡易辰在專題中使用的開發語言是 Unity。
verdict: 1
reason: The context explicitly states that the language used is Unity.

statement: 遊戲邏輯是用 C#。
verdict: 1
reason: The context mentions that the game logic is implemented using C#.

statement: 資料庫仍然使用 MySQL。
verdict: 1
reason: The context confirms that the database is still using MySQL.

statement: Unity 擁有自己的圖形化介面。
verdict: 1
reason: The context states that Unity has its own graphical interface.

statement: Unity 的圖形化介面使得開發過程更加直觀。
verdict: 0
reason: While the context mentions that Unity has a graphical interface, it does not explicitly state that it makes the development process more intuitive.

statement: 使用相同的資料庫 MySQL 可以降低學習成本和時間。
verdict: 0
reason: The context does not provide information about the learning costs or time related to using the same database, so this cannot be inferred.

statement: 蔡易辰並未詳細提及具體缺點。
verdict: 1
reason: The context mentions that there are difficulties and problems encountered, but does not provide specific detai

Evaluating:  30%|███       | 61/200 [00:31<01:18,  1.78it/s]


statement: 彭文暘在系統開發中遇到的問題主要有兩個。
verdict: 0
reason: The context does not specify a number of problems encountered by 彭文暘, only discusses issues related to user needs and surveys.

statement: 第一個問題是不清楚使用者的需求。
verdict: 1
reason: The context mentions that sometimes the team is unclear about user needs, which aligns with this statement.

statement: 團隊有時候不清楚使用者到底需要哪些功能。
verdict: 1
reason: This statement is directly supported by the context, which states that the team sometimes does not know what features users need.

statement: 因此，團隊需要定期調查使用者的需求。
verdict: 1
reason: The context suggests that regular surveys are necessary to understand user needs, supporting this statement.

statement: 團隊需要在組內討論哪些功能真的需要，哪些功能可以省略。
verdict: 1
reason: The context mentions the need for team discussions about necessary and unnecessary features, validating this statement.

statement: 第二個問題是需求調查不完善。
verdict: 1
reason: The context indicates that the initial user needs survey was not thorough, which supports this stat

Evaluating:  31%|███       | 62/200 [00:33<01:45,  1.31it/s]


statement: 根據吳少宇的經驗，在開會的時候，會檢查每位組員是否完成上禮拜分配的工作。
verdict: 1
reason: 吳少宇提到開會最重要的是看大家有沒有做到上禮拜分配好的工作，因此這個陳述可以直接推斷。

statement: 檢查主要是透過討論來了解進度。
verdict: 0
reason: 雖然討論是開會的一部分，但具體的檢查方式並未明確提到，因此無法直接推斷。

statement: 如果有人沒辦法趕上進度，討論會讓大家知道這個情況。
verdict: 0
reason: 吳少宇提到如果有人無法完成工作，會給予道德上的譴責，但並未提到討論會讓大家知道這個情況，因此無法直接推斷。

statement: 這個情況最終會造成該組員的成績影響。
verdict: 1
reason: 吳少宇提到期末時會根據工作表給予該組員5%到10%的成績，因此這個陳述可以直接推斷。

statement: 開會的目的就是看大家有沒有完成預先分配的任務。
verdict: 1
reason: 吳少宇明確提到開會最重要的就是看大家有沒有做到上禮拜分配好的工作，因此這個陳述可以直接推斷。


Evaluating:  32%|███▏      | 64/200 [00:33<01:07,  2.00it/s]


statement: 在系統分析與設計課程中，學生在專案開發過程中主要遇到的問題包括技術上的困難、組員之間的協調與分工、以及溝通不暢等。
verdict: 1
reason: 根據多位學生的經驗，提到的問題如技術困難、組員協調和溝通不暢都是在專案開發過程中遇到的，因此這個陳述可以直接推斷。

statement: 根據吳少宇的經驗，學生在技術上會有比較大的困難。
verdict: 1
reason: 吳少宇明確提到在專案開發中遇到的技術困難，因此這個陳述是可以直接推斷的。

statement: 學校的課程教的內容可能不足以支持學生完成一個完整的系統。
verdict: 1
reason: 吳少宇提到學校的課程教的內容遠遠不夠，因此這個陳述是可以直接推斷的。

statement: 林芯緹與蘇彥碩強調了組員的選擇與分工協調的重要性。
verdict: 1
reason: 林芯緹和蘇彥碩的經驗中提到組員的選擇和分工協調是關鍵，因此這個陳述是可以直接推斷的。

statement: 若分工不當則可能會影響專案進度。
verdict: 1
reason: 根據學長姐的經驗，分工不當會影響專案進度，因此這個陳述是可以直接推斷的。

statement: 朱浩偉指出，自身技術及程度不足也是一大挑戰。
verdict: 1
reason: 朱浩偉明確提到自身技術及程度不夠是挑戰，因此這個陳述是可以直接推斷的。

statement: 學生必須不斷精進自我能力以趕上預期的開發成果。
verdict: 1
reason: 朱浩偉提到必須不斷精進自我能力，因此這個陳述是可以直接推斷的。

statement: 學生反映學校的課程在技術支持方面是遠遠不夠的。
verdict: 1
reason: 吳少宇提到學校的課程教的內容遠遠不夠，因此這個陳述是可以直接推斷的。

statement: 這使得學生在專案開發過程中需要自我加強並尋找額外的資源來克服困難。
verdict: 1
reason: 根據多位學生的經驗，提到需要自我加強和尋找額外資源來克服困難，因此這個陳述是可以直接推斷的。

statement: 根據辛語柔的經驗，換 vscode 的顏色可以讓人感到更有動力。
verdict: 1
reason: 辛語柔提到換顏色讓她在打程式時感到很爽，這暗示了這樣的改變能提升動力。



Exception raised in Job[61]: InstructorRetryException(<failed_attempts>

<generation number="1">
<exception>
    Error code: 429 - {'error': {'message': 'Rate limit reached for gpt-4o-mini in organization org-R5IFxMmFpVw4rgKpDM8hcA83 on tokens per min (TPM): Limit 200000, Used 200000, Requested 1249. Please try again in 374ms. Visit https://platform.openai.com/account/rate-limits to learn more.', 'type': 'tokens', 'param': None, 'code': 'rate_limit_exceeded'}}
</exception>
<completion>
    None
</completion>
</generation>

<generation number="2">
<exception>
    Error code: 429 - {'error': {'message': 'Rate limit reached for gpt-4o-mini in organization org-R5IFxMmFpVw4rgKpDM8hcA83 on tokens per min (TPM): Limit 200000, Used 200000, Requested 1249. Please try again in 374ms. Visit https://platform.openai.com/account/rate-limits to learn more.', 'type': 'tokens', 'param': None, 'code': 'rate_limit_exceeded'}}
</exception>
<completion>
    None
</completion>
</generation>

<generation num


statement: 根據蔡易辰的經驗，蔡易辰建議不要只用 MySQL。
verdict: 1
reason: 蔡易辰提到強烈建議不要只用 MySQL，這是直接從上下文中可以推斷的。

statement: 蔡易辰的建議是因為在 SA 階段仍有更多時間可以學習其他程式語言和資料庫。
verdict: 1
reason: 上下文中提到在 SA 階段有更多時間可以學習其他程式語言，這支持了蔡易辰的建議。

statement: 如果專題需要使用另一種資料庫，可能會需要花時間從頭學習。
verdict: 1
reason: 上下文中提到如果專題要用另一個語言開發的時候，會需要花時間從頭去學，這可以推斷出對於資料庫也是如此。

statement: 資料中未提及其他可選擇的資料庫的具體名稱。
verdict: 1
reason: 上下文中並未提及具體的其他資料庫名稱，因此這一陳述是正確的。

statement: 除了 MySQL，常見的資料庫還包括 PostgreSQL、MongoDB、SQLite 等。
verdict: 0
reason: 上下文中並未提及這些資料庫的名稱，因此無法直接推斷出這一陳述的正確性。

statement: 可以根據專案需求選擇適合的資料庫。
verdict: 0
reason: 上下文中提到可以學習其他程式語言和資料庫，但並未明確提到根據專案需求選擇資料庫，因此這一陳述無法直接推斷。


Exception raised in Job[117]: InstructorRetryException(<failed_attempts>

<generation number="1">
<exception>
    Error code: 429 - {'error': {'message': 'Rate limit reached for gpt-4o-mini in organization org-R5IFxMmFpVw4rgKpDM8hcA83 on tokens per min (TPM): Limit 200000, Used 200000, Requested 1308. Please try again in 392ms. Visit https://platform.openai.com/account/rate-limits to learn more.', 'type': 'tokens', 'param': None, 'code': 'rate_limit_exceeded'}}
</exception>
<completion>
    None
</completion>
</generation>

<generation number="2">
<exception>
    Error code: 429 - {'error': {'message': 'Rate limit reached for gpt-4o-mini in organization org-R5IFxMmFpVw4rgKpDM8hcA83 on requests per min (RPM): Limit 500, Used 500, Requested 1. Please try again in 120ms. Visit https://platform.openai.com/account/rate-limits to learn more.', 'type': 'requests', 'param': None, 'code': 'rate_limit_exceeded'}}
</exception>
<completion>
    None
</completion>
</generation>

<generation number=


statement: The speaker cannot find relevant information.
verdict: 0
reason: The context provides detailed information about the experiences and opinions regarding Firebase and React, indicating that the speaker has relevant information.

statement: The speaker suggests asking about specific contexts in which MySQL, MongoDB, and Firebase are used.
verdict: 0
reason: The context does not mention any suggestion to ask about specific contexts for MySQL, MongoDB, and Firebase. It focuses on the experiences with Firebase and React.

statement: The speaker also suggests inquiring about the advantages and disadvantages of MySQL, MongoDB, and Firebase.
verdict: 0
reason: While the context discusses the advantages and disadvantages of Firebase and React, it does not suggest inquiring about MySQL or MongoDB specifically.


Exception raised in Job[138]: InstructorRetryException(<failed_attempts>

<generation number="1">
<exception>
    Error code: 429 - {'error': {'message': 'Rate limit reached for gpt-4o-mini in organization org-R5IFxMmFpVw4rgKpDM8hcA83 on tokens per min (TPM): Limit 200000, Used 199043, Requested 1869. Please try again in 273ms. Visit https://platform.openai.com/account/rate-limits to learn more.', 'type': 'tokens', 'param': None, 'code': 'rate_limit_exceeded'}}
</exception>
<completion>
    None
</completion>
</generation>

<generation number="2">
<exception>
    Error code: 429 - {'error': {'message': 'Rate limit reached for gpt-4o-mini in organization org-R5IFxMmFpVw4rgKpDM8hcA83 on requests per min (RPM): Limit 500, Used 500, Requested 1. Please try again in 120ms. Visit https://platform.openai.com/account/rate-limits to learn more.', 'type': 'requests', 'param': None, 'code': 'rate_limit_exceeded'}}
</exception>
<completion>
    None
</completion>
</generation>

<generation number=


statement: 根據朱浩偉的經驗，朱浩偉在專題製作中建議要安排一個計畫。
verdict: 1
reason: 朱浩偉的經驗中提到要安排計畫以確保小組進度，因此這個陳述可以直接推斷。

statement: 安排計畫可以確保小組進度能夠跟著計劃進行。
verdict: 1
reason: 根據朱浩偉的經驗，安排計畫的確是為了確保小組進度，因此這個陳述是正確的。

statement: 朱浩偉建議把握每個可以寫小專案的機會。
verdict: 1
reason: 朱浩偉的經驗中提到要把握每一個可以寫小專案的機會，因此這個陳述是正確的。

statement: 釐清系統開發對使用者的價值會讓專題製作更順利。
verdict: 1
reason: 朱浩偉提到釐清系統開發對使用者的價值會使專題製作更順利，因此這個陳述是正確的。

statement: 至於時間投入方面，初期每週可能需要10-15小時。
verdict: 1
reason: 朱浩偉提到初期每週可能需要10-15小時的時間投入，因此這個陳述是正確的。

statement: 初期的時間主要用於研究和規劃。
verdict: 1
reason: 根據朱浩偉的經驗，初期的時間確實主要用於研究和規劃，因此這個陳述是正確的。

statement: 在進入開發和整合階段，接近死線或與產學合作方有審查時，每週投入的時間會增加到20-30小時甚至更多。
verdict: 1
reason: 朱浩偉提到在開發和整合階段，時間投入會增加到20-30小時甚至更多，因此這個陳述是正確的。


Evaluating: 100%|█████████▉| 199/200 [01:54<00:01,  1.16s/it]


statement: 林芯緹和蘇彥碩建議在共同開發時要事先溝通好程式的命名規則。
verdict: 1
reason: 根據林芯緹、蘇彥碩的經驗，確實提到要溝通好程式之命名規則，因此這個陳述可以直接推斷。

statement: 程式的命名規則最好以功能來命名。
verdict: 1
reason: 學姐建議以功能命名，這與陳述一致，因此可以直接推斷。

statement: 這樣可以讓所有人以及未來的自己都清楚程式碼的意義。
verdict: 1
reason: 文中提到可以將自己的邏輯想法寫在註解，讓大家和未來的自己都清楚程式碼的意義，因此這個陳述可以直接推斷。

statement: 另外，建議在開發過程中定期舉行「debug 大會」來追蹤組員的進度。
verdict: 1
reason: 文中提到偶爾開個debug 大會，這與陳述一致，因此可以直接推斷。

statement: 在「debug 大會」中，應該適時關心每位組員的狀況。
verdict: 1
reason: 文中提到並適時關心每一個組員進度，這與陳述一致，因此可以直接推斷。

statement: 這樣可以確保大家都在同一頁。
verdict: 1
reason: 雖然文中沒有明確提到這一點，但關心組員的狀況有助於確保大家在同一頁，因此可以推斷出這個陳述的合理性。

statement: 提高整體合作的效率是最終目標。
verdict: 1
reason: 文中提到要分工合作才能完成整個Project，這暗示了提高合作效率的重要性，因此可以推斷出這個陳述的合理性。


Evaluating: 100%|██████████| 200/200 [01:55<00:00,  1.73it/s]


statement: 根據蔡易辰的經驗，使用MySQL的優點是因為上學期（大二上）才剛學過。
verdict: 1
reason: 蔡易辰提到使用MySQL的原因是因為上學期剛學過，因此這個陳述可以直接推斷。

statement: 對於學生來說，使用MySQL比較熟悉。
verdict: 0
reason: 雖然蔡易辰提到使用MySQL的原因，但並未明確指出學生對MySQL的熟悉程度，因此無法直接推斷。

statement: 學生可以快速上手使用MySQL。
verdict: 0
reason: 蔡易辰提到上學期剛學過MySQL，但並未提到學生能否快速上手，因此無法直接推斷。

statement: 蔡易辰強烈建議不要只用MySQL這個資料庫。
verdict: 1
reason: 蔡易辰在上下文中明確提到不建議只用MySQL，因此這個陳述可以直接推斷。

statement: 在SA階段還有更多時間學習其他技術。
verdict: 1
reason: 蔡易辰提到在SA階段有更多時間學習其他程式語言，因此這個陳述可以直接推斷。

statement: 蔡易辰並未明確提到MySQL的缺點。
verdict: 1
reason: 蔡易辰在上下文中並未提到MySQL的缺點，因此這個陳述可以直接推斷。

statement: 如果專題要用另一個語言開發，可能會需要重新學習。
verdict: 1
reason: 蔡易辰提到如果專題要用另一個語言開發，會需要花時間從頭學習，因此這個陳述可以直接推斷。

statement: 這意味著若選擇MySQL而未掌握其他資料庫，轉換時可能會影響專題的進度。
verdict: 0
reason: 雖然蔡易辰提到需要學習其他技術，但並未明確提到選擇MySQL會影響專題進度，因此無法直接推斷。

statement: 目前沒有彭文暘的相關經驗資料。
verdict: 1
reason: 上下文中並未提到彭文暘的經驗，因此這個陳述可以直接推斷。

statement: 無法提供彭文暘的觀點。
verdict: 1
reason: 由於上下文中沒有提到彭文暘的觀點，因此這個陱述可以直接推斷。
{'faithfulness': 0.8351, 'context_precision': 0.9500, 'context_recall': 0.9

In [30]:
from ragas import evaluate
from ragas.metrics import (
    AnswerCorrectness,
    ContextPrecision,
    ContextRecall,
    Faithfulness,
    AnswerRelevancy,
)
import openai
from ragas.llms import llm_factory
from ragas.embeddings import embedding_factory
from ragas.embeddings import GoogleEmbeddings
from google import genai
from dotenv import load_dotenv
from ragas.embeddings import LangchainEmbeddingsWrapper
from langchain_openai import OpenAIEmbeddings

load_dotenv()

import pandas as pd
import ast
from ragas import EvaluationDataset

# 讀取 CSV
df = pd.read_csv("ragas_eval_with_response_2026-05-22.csv", encoding="utf-8-sig")

# 還原 list 欄位
df["retrieved_contexts"] = df["retrieved_contexts"].apply(ast.literal_eval)

# 欄位重新命名
ragas_df = df[["user_input", "retrieved_contexts", "response", "reference"]]

rag_results = EvaluationDataset.from_pandas(ragas_df)


# 用 AsyncOpenAI
async_client = openai.AsyncOpenAI()

evaluator_llm = llm_factory("gpt-4o-mini", provider="openai", client=async_client,max_tokens=4096)
ragas_embeddings = LangchainEmbeddingsWrapper(
    OpenAIEmbeddings(model="text-embedding-3-small")
)

answer_relevancy = AnswerRelevancy(llm=evaluator_llm, embeddings=ragas_embeddings)
answer_relevancy.question_generation.instruction = """根據給定的答案，反推出最有可能對應的問題，並判斷該答案是否為模糊回答。
如果答案是模糊的，noncommittal 給 1；如果答案是明確的，noncommittal 給 0。
模糊回答是指那些迴避、含糊或不明確的回答。
例如，「我不知道」或「我不確定」都屬於模糊回答。
請用繁體中文產生問題。"""

scores = evaluate(
    dataset=rag_results,
    metrics=[
        Faithfulness(llm=evaluator_llm),
        ContextPrecision(llm=evaluator_llm),
        ContextRecall(llm=evaluator_llm),
        answer_relevancy
        # AnswerCorrectness(llm=evaluator_llm),
    ],
)


print(scores)

C:\Users\User\AppData\Local\Temp\ipykernel_11952\2593691224.py:2: DeprecationWarning: Importing AnswerCorrectness from 'ragas.metrics' is deprecated and will be removed in v1.0. Please use 'ragas.metrics.collections' instead. Example: from ragas.metrics.collections import AnswerCorrectness
  from ragas.metrics import (
C:\Users\User\AppData\Local\Temp\ipykernel_11952\2593691224.py:2: DeprecationWarning: Importing ContextPrecision from 'ragas.metrics' is deprecated and will be removed in v1.0. Please use 'ragas.metrics.collections' instead. Example: from ragas.metrics.collections import ContextPrecision
  from ragas.metrics import (
C:\Users\User\AppData\Local\Temp\ipykernel_11952\2593691224.py:2: DeprecationWarning: Importing ContextRecall from 'ragas.metrics' is deprecated and will be removed in v1.0. Please use 'ragas.metrics.collections' instead. Example: from ragas.metrics.collections import ContextRecall
  from ragas.metrics import (
C:\Users\User\AppData\Local\Temp\ipykernel_1195

{'faithfulness': 0.7917, 'context_precision': 0.9615, 'context_recall': 0.9038, 'answer_relevancy': 0.9091}


ragas_eval_with_response_v3
ragas_testset_v2
{'context_precision': 0.9833, 'context_recall': 0.9000, 'answer_relevancy': 0.6260, 'answer_correctness': 0.7897}

3.1
{'context_precision': 0.9833, 'context_recall': 0.9000, 'answer_relevancy': 0.6260, 'answer_correctness': 0.7897}

"ragas_eval_with_response_2026-05-22.csv"

{
    "faithfulness": 0.7917,
    "context_precision": 0.9615,
    "context_recall": 0.9038,
    "answer_relevancy": 0.9091,
}

ragas_eval_with_response_v3
{'faithfulness': 0.8351, 'context_precision': 0.9500, 'context_recall': 0.9333, 'answer_relevancy': 0.7727}